### PR TITLE
[timeseries] Train multiple models for multi-window backtesting 

### PR DESCRIPTION
**Pull request by shchur**
Current implementation of multi-window backtesting in AG-TS 1/ provides an overly pessimistic estimate of the model performance and 2/ results in models not being trained on the most recent data.

This PR changes the multi-window backtesting logic & addresses this problem.

<details><summary>Description of the proposed approach</summary> 

Assume the dataset contains a single time series:
```
[x x x x x x x x x x x x x x]
```
**Current approach** (master branch)  train single model & evaluate on multiple windows
```
[x x x x x x x x] - train data

[x x x x x x x x x x x ? ? ?]  - val_data (last 3 values used for eval)
[x x x x x x x x ? ? ?]
```

**Proposed approach** (this PR) train a separate model for each window

Model 1
```
[x x x x x x x x x x x] - train data
[x x x x x x x x x x x ? ? ?]  - val data
```
Model 2
```
[x x x x x x x x] - train data
[x x x x x x x x ? ? ?]  - val data
```
</details>

*Description of changes:*
- Introduce parameter `num_val_windows: int = 1` to `TimeSeriesPredictor.fit` that controls how many windows are used for multi-window backtesting.
- Introduce a meta-model `MultiWindowBacktestingModel` that trains multiple copies of a base model under the hood and uses different windows to evaluate the performance.
- Deprecate the `validation_splitter` argument passed to `TimeSeriesPredictor` as this functionality is now controlled via `num_val_windows`.
- Add stricter check to `TimeSeriesPredictor` ensure that the training time series are long enough.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py [hunk 1]
```diff
@@ -379,6 +379,13 @@ def score_and_cache_oof(
         if store_val_score:
             self.val_score = self._score_with_predictions(val_data, self._oof_predictions)
 
+    def _update_hpo_train_fn_kwargs(self, train_fn_kwargs: dict) -> dict:
```
Comment (canerturkmen) [inline timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py:382]: Is it more natural to make this `_get_hpo_train_fn_kwargs` instead of `_update`?

File: timeseries/src/autogluon/timeseries/predictor.py [hunk 2]
```diff
@@ -270,6 +290,7 @@ def fit(
         presets: Optional[str] = None,
         hyperparameters: Dict[Union[str, Type], Any] = None,
         hyperparameter_tune_kwargs: Optional[Union[str, Dict]] = None,
+        num_val_windows: int = 1,
```
Comment (canerturkmen) [inline timeseries/src/autogluon/timeseries/predictor.py:293]: My main issue with this PR is that we appear to be forcing _all_ models to be essentially bagged models with a single parameter. For validation splitter, I think this was OK since this choice was primarily about model selection and validation. Now, it seems to be primarily about training. For example, I could be much more OK doing this for some cheap models and not other more expensive ones?

**Review by canerturkmen — COMMENTED** [review]
Review WIP. Some comments so far.

File: timeseries/tests/unittests/test_trainer.py [hunk 3]
```diff
@@ -38,7 +38,6 @@ def trained_trainers():
         )
         trainer.fit(
             train_data=DUMMY_TS_DATAFRAME,
-            val_data=DUMMY_TS_DATAFRAME,
```
Comment (canerturkmen) [inline timeseries/tests/unittests/test_trainer.py:41]: why must we remove val_data here?

**Review by canerturkmen — COMMENTED** [review]
this was a great read. thanks! 🚀

File: timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py [hunk 4]
```diff
@@ -817,3 +816,45 @@ def get_model_inputs_for_scoring(
         else:
             known_covariates = None
         return past_data, known_covariates
+
+    def train_test_split(
+        self,
+        prediction_length: int,
+        window_idx: int = 0,
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py]: Changed this to an optional `end_index` parameter. If `end_index` is given, then all time series are trimmed up to `end_idx` before the split.

File: timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        self.val_score = np.mean([model.val_score for model in trained_models])
+
+    def get_info(self) -> dict:
+        info = super().get_info()
+        info["info_per_val_window"] = self.info_per_val_window
+        return info
+
+    def get_child_model(self, window_idx: int) -> AbstractTimeSeriesModel:
+        model = copy.deepcopy(self.model_base)
+        model.set_contexts(self.path + f"W{window_idx}" + os.sep)
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py]: The model will be re-initialized when `fit` is called (see https://github.com/autogluon/autogluon/blob/master/core/src/autogluon/core/models/abstract/abstract_model.py#L748), so I think it's redundant.

File: timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        return self.most_recent_model.predict(data, known_covariates=known_covariates, **kwargs)
+
+    def score_and_cache_oof(
+        self,
+        val_data: TimeSeriesDataFrame,
+        store_val_score: bool = False,
+        store_predict_time: bool = False,
+    ) -> None:
+        # self.val_score, self.predict_time, self._oof_predictions already saved during _fit()
+        assert self._oof_predictions is not None
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py:153]: Agreed, I removed the assertion from the parent (`AbstractTimeSeriesModel`) but kept it here.

This will produce the same effect for both `MultiWindowBacktestingModel` and any other model:
- multi-window case: assertion will trigger if the model hasn't been fit before
- usual model: assertion will trigger because the model cannot predict (since it hasn't been fit yet)

File: timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                    "val_score": model.val_score,
+                    "predict_time": model.predict_time,
+                }
+            )
+
+        # Only the model trained on most recent data is saved & used for prediction
+        self.most_recent_model = trained_models[0]
+        self.most_recent_model_path = trained_models[0].path
+        self.predict_time = self.most_recent_model.predict_time
+        self.fit_time = time.time() - global_fit_start_time - self.predict_time
```
[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py:122]: Since in Tabular prediction is done with an ensemble consisting of all trained models, there `fit_time = sum(model.fit_time for model in bagged_models)` and `predict_time = sum(model.predict_time for model in bagged_models)`.

My idea here was to satisfy the following conditions:
- time that we spent fitting & scoring the model (excluding metric computation time) = `fit_time + predict_time` 
- time that it actually takes to predict with `MultiWindowBacktestingModel` = `predict_time`

Does this sound reasonable?

[the remark this answers is not in the document]
Comment (Innixma) [reply timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py:122]: "fit_time = sum(model.fit_time for model in bagged_models)" Not actually the case in Tabular. This used to be the logic, but after we added parallel bagging this is no longer the logic, instead it is the end-to-end wall-clock time of the fit call of the bagged ensemble, although you can check the code for the nuance.

[the remark this answers is not in the document]
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py:122]: Thanks, Nick! Here we have sequential fitting of folds, so I think what we report is equal to both the end-to-end wall-clock time and the sum of fit_times for individual models. I agree that if we implement parallel fold fitting here, we should report the wall-clock time, like in ParallelFoldFittingStrategy in Tabular.

**Review by canerturkmen — APPROVED** [review]
